### PR TITLE
Rule S3971: 'GC.SuppressFinalize' should not be called

### LINKED ISSUE #210: Rule S3971: "GC.SuppressFinalize" should not be called
Implements [RSPEC-3971: "GC.SuppressFinalize" should not be called](https://jira.sonarsource.com/browse/RSPEC-3971)

Contributes to [MMF-820: SonarC#: fill the gap with FXCop (Step 4)](https://jira.sonarsource.com/browse/MMF-820)

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #210

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalize.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System.Collections.Immutable;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotCallGCSuppressFinalize : SonarDiagnosticAnalyzer
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalize.cs]: MINOR Rename class 'DoNotCallGCSuppressFinalize' to match camel case naming rules, consider using 'DoNotCallGcSuppressFinalize'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS101)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalize.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        internal const string DiagnosticId = "S3971";
+        private const string MessageFormat = "Do not call 'GC.SuppressFinalize'.";
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalize.cs]: Do you think that my message on gg is a good idea? If yes, this is a nice opportunity to change the message :)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalize.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(ReportIfInvalidMethodIsCalled,
+                SyntaxKind.InvocationExpression);
+        }
+
+        private void ReportIfInvalidMethodIsCalled(SyntaxNodeAnalysisContext analysisContext)
+        {
+            var invocation = (InvocationExpressionSyntax)analysisContext.Node;
+
+            var identifier = GetMethodCallIdentifier(invocation);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalize.cs]: I know you are trying to check the syntax first and then obtain the symbols for performance reasons, but the code will be much simpler if you just obtain the symbol and then do the check on it (preferably in a KnownMethods extension method):

    var symbol = analysisContext.SemanticModel.GetSymbolInfo(invocation).Symbol;
    if (symbol.IsGcSuppressFinalize())
    {
        var methodDeclarationSymbol = ...
        if (methodDeclarationSymbol.IsIDisposableDispose())
        {
            // report
        }
    }

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodShouldNotBeCalled.cs [hunk 4]
```diff
@@ -58,12 +58,8 @@ private void ReportIfInvalidMethodIsCalled(SyntaxNodeAnalysisContext analysisCon
             }
 
             var methodCallSymbol = analysisContext.SemanticModel.GetSymbolInfo(identifier.Value.Parent);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodShouldNotBeCalled.cs]: Why not use MethodShouldNotBeCalled as base class?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalize.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            if (methodCallSymbol.Symbol == null ||
+                !methodCallSymbol.Symbol.ContainingType.ConstructedFrom.Is(KnownType.System_GC))
+            {
+                return;
+            }
+
+            var methodDeclaration = invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>();
+            if (methodDeclaration == null)
+            {
+                analysisContext.ReportDiagnostic(Diagnostic.Create(rule, identifier.Value.GetLocation()));
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalize.cs]: I guess this is for properties and events too, may be you should put a test that ensures that some day I will not delete this strange block :)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalizeMethod.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System.Collections.Generic;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotCallGCSuppressFinalizeMethod : DoNotCallMethodsBase
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCSuppressFinalizeMethod.cs]: MINOR Rename class 'DoNotCallGCSuppressFinalizeMethod' to match camel case naming rules, consider using 'DoNotCallGcSuppressFinalizeMethod'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS101)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCCollectMethod.cs [hunk 7]
```diff
@@ -1,47 +1,46 @@
...
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System.Collections.Generic;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotCallGCCollectMethod : DoNotCallMethodsBase
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallGCCollectMethod.cs]: MINOR Rename class 'DoNotCallGCCollectMethod' to match camel case naming rules, consider using 'DoNotCallGcCollectMethod'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS101)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Pull request merged**